javascript/resources/tools/autobuild.sh

```diff
@@ -4,12 +4,33 @@ set -eu
 
 jvm_args=-Xss16m
 
-# If CODEQL_RAM is set, use half for Java and half for TS.
+# If CODEQL_RAM is set, try to automatically calculate how much memory is available for TS and Java
+# if no explicit values are set for them.
 if [ -n "${CODEQL_RAM:-}" ] ; then
     half_ram="$(( CODEQL_RAM / 2 ))"
-    LGTM_TYPESCRIPT_RAM="$half_ram"
+
+    # If either CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY or CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY is set,
+    # calculate the other by subtracting from CODEQL_RAM. If neither is set, then use
+    # half of the available CODEQL_RAM for each.
+    if [ -n "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY:-}" ] && [ -z "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY}" ] ; then
+        CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY="$(( $CODEQL_RAM - $CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY ))"
+    elif [ -n "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY:-}" ] && [ -z "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY}" ] ; then
+        CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY="$(( $CODEQL_RAM - $CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY ))"
+    elif [ -z "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY}" ] && [ -z "${CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_TS_MEMORY}" ] ; then
```

Comment:
I tried to do a DCA experiment with this change, but I got a big error around here.  
DCA link: https://github.com/codeql-dca-runners/codeql-dca-worker_javascript/actions/runs/9890760504/job/27319703621 

I got this error: 
_[2024-07-11 11:43:40] [build-stderr] /home/runner/work/_temp/b8b25efe-4bd4-4a9f-a121-2f257e642df2/codeql/javascript/tools/autobuild.sh: 19: CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY: parameter not set_

I'm not sure why, as I'm not that much into shell scripts, but it seems like something that needs fixing.  

You can ask in #codeql-dca for how to run such experiments, and you should probably also do a test on windows.

Reply:
Not sure what you mean about DCA - does this need a special experiment? In any case, I also don't know much about bash, but with the power of copilot I learned what some of this syntax means and I suspect it's because `{CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY}` should be `{CODEQL_EXTRACTOR_JAVASCRIPT_OPTION_JVM_MAX_MEMORY:-}` on line 19. And maybe that `:-` should be used for some of the other parameter expansions in this part of the code.

Reply:
> Not sure what you mean about DCA  - does this need a special experiment?

No, DCA does not need a special environment, it's just the tool we use to simulate how CodeQL is used in production.  
I get the same error when I try your change on my own laptop, so it definitely needs to be fixed. (When running `codeql database create -l=javascript ...`).  

I don't know much about bash either, so I can't really help you with how to fix it.  
But I would recommend building the CodeQL CLI locally and trying to run `codeql database create ...` to test that the change doesn't crash everything.  
And I recommend using the same locally build CLI to test that the change actually works. 

It should also be tested on Windows.  
If you don't have access to a Windows machine, then we can probably use DCA to test on Windows. We can get back to that when it works on Unix.